ls1intum/Artemis
# [Exam] Filtering for exam score statistics

### Checklist

### Motivation and Context
Introduces filters for the exam score page statistics, which enables instructors to get a more reasonable picture of the exam statistics.

### Description

- [x] Use `ChangeDetectionStrategy.OnPush` for the exam scores component to improve the user experience
- [x] Add filters for the calculation of exercise statistics: One filters out all un-submitted exams and the other filters out all exercises with only empty submissions. Filters only apply to the first table and don't affect the export of results or the student results table
- [x] Calculate the average scores and the number of participants for each exercise and exerciseGroup while taking the filters into account. This calculation is done on the client to avoid multiple REST requests
- [x] Refactor and extend parts of the `ExamScoresDTO`. Contains now the total number of participants per exercise and exercise group. Each `ExerciseResult` stores a flag now indicating whether non empty submissions exist for this exercise.
- [x] Remove now obsolete calculations on server-side and improve performance in general
- [x] Improve the integration test 

### Steps for Testing


1. Create an exam with all exercise types, and multiple exercises per exercise group. You can also tryout optional exercise groups.
2. Conduct the exam with multiple users. Do not work on some exercise and do not submit the exam for some users. Note down what you did with which user.
3. Assess the exam
4. Check the exam scores page and play around with filters
-> Check that the statistics are correctly calculated

### Screenshots


![newExamScoresPage](https://user-images.githubusercontent.com/16407766/89447331-1b18f080-d756-11ea-893b-29757e4055c9.PNG)
